### ISSUE TITLE
Security concern with HMR?

### ISSUE THREAD

**Issue by Jojoshua** [sveltejs/kit#7191]
### Describe the bug

Normally when you try to import and use $env/dynamic/private in a public context an error will be thrown.

`Cannot import $env/dynamic/private into public-facing code:`

However there is a way you can temporarily get around this behavior by starting the app and refreshing via HMR.

### Reproduction

1. Clone https://github.com/Jojoshua/sveltekit-double-mount/tree/PrivateENV
2. Start with the dev script
3. Uncomment lines 4 and 5 from src/routes/+page.svelte and save
4. Observe the private_env is printed in the console

### System Info

```shell
System:
    OS: Windows 10 10.0.19043
    CPU: (16) x64 AMD Ryzen 7 1700X Eight-Core Processor
    Memory: 4.79 GB / 15.93 GB
  Binaries:
    Node: 18.9.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.18 - ~\AppData\Roaming\npm\yarn.CMD
    npm: 8.4.1 - ~\AppData\Roaming\npm\npm.CMD
  Browsers:
    Chrome: 106.0.5249.103
    Edge: Spartan (44.19041.1266.0), Chromium (106.0.1370.37)
    Internet Explorer: 11.0.19041.1566
  npmPackages:
    @sveltejs/adapter-auto: 1.0.0-next.80 => 1.0.0-next.80
    @sveltejs/kit: 1.0.0-next.504 => 1.0.0-next.504
    @sveltejs/package: 1.0.0-next.5 => 1.0.0-next.5
    svelte: ^3.50.1 => 3.50.1
    vite: ^3.1.3 => 3.1.3
```

### Severity

annoyance

**Comment by dummdidumm** [comment]
I wouldn't call this a security concern since this will be caught when building for production. Also I'm not sure how much we can do here given the HMR trickeries, but I agree it's raising an eyebrow.

**Comment by Jojoshua** [comment]
Yes I'm not sure what the implications are so wanted to bring this to the team to review.

I'm not much concerned with local development but curious about online development environments like stack blitz, etc.

**Comment by tcc-sejohnson** [comment]
This isn't a security concern, as it only happens in dev. I don't believe there's any way for us to get around it, but I'm totally open to suggestions if someone can think of a way. I'm not sure what implications you're thinking of in online dev environments like StackBlitz... you'd have to have the environment variables exposed to include them in a StackBlitz repo anyway...?

**Comment by dominikg** [comment]
3rding  the "not a security concern". Additional arguments
- You really should not use production secrets on your local machine, that's what `.env.local` is for
- The dev-server doesn't listen on external interfaces by default
- The dev-server must not be used in production

You'd have to take a series of steps in the wrong direction before this could theoretically be a problem

As for warning the user during dev as opposed to build, this would need to be happen in either `handleHotUpdate` or `transform` hook by throwing an error. Right now the check happens during writeBundle. https://github.com/sveltejs/kit/blob/master/packages/kit/src/exports/vite/index.js#L358

**Comment by Jojoshua** [comment]
Good info @dominikg . In my case I was testing how I could make configuration replacements (for a library) at compile time by using `import { env as public_env } from '$env/dynamic/public'` but ended up testing dynamic/private also.

I discovered the replacements don't actually work when packaging (similar to https://github.com/sveltejs/kit/issues/3480) so I moved away from env entirely. 

I don' t currently have an exploit to take advantage of this behavior but just had some general follow-on thoughts...

- Developers are increasingly the targets of cyber-attacks & the source of breaches
- Once a developer finds a way of "making it work" they will stick with it
- An attacker will often use minor "quirks" as an advantage to escalate and create a more sophisticated attack than otherwise possible

I'm happy for a member to close this issue whenever it makes sense. Thanks for the quick review guys!

**Comment by tcc-sejohnson** [comment]
> - Developers are increasingly the targets of cyber-attacks & the source of breaches
> - Once a developer finds a way of "making it work" they will stick with it
> - An attacker will often use minor "quirks" as an advantage to escalate and create a more sophisticated attack than otherwise possible

I don't see any way this could be used to "expose" a private environment variable. If someone has the level of access required to

- Visit a specific route (on the dev server)
- Trigger HMR
- Parse the HTML for a specific value

They could just... read the environment variable from the environment. Or screw you over in any number of other ways. 

Ideally, this wouldn't happen, but I really don't know a way around it, and I don't think it's ever going to cause an issue.

**Comment by dummdidumm** [comment]
Closing as this isn't a security concern - as pointed out, dev shouldn't be used in prod environments, and if someone has access to your dev environment, worse things will happen than having access to the secret in the running app.
